drivers/i2s/i2s_mcux_flexcomm.c

```diff
@@ -0,0 +1,940 @@
+					break;
+				}
+				LOG_ERR("buffer alloc from slab %p err %d",
+					stream->cfg.mem_slab, ret);
+				i2s_mcux_rx_stream_disable(dev, false);
+				stream->state = I2S_STATE_ERROR;
+			} else {
+				const struct i2s_mcux_config *cfg = dev->config;
+				I2S_Type *base = cfg->base;
+
+				dma_reload(stream->dev_dma, stream->channel,
+					   (uint32_t)&base->FIFORD, (uint32_t)buffer,
+					   stream->cfg.block_size);
+				dma_start(stream->dev_dma, stream->channel);
+
+				/* put buffer in input queue */
+				ret = k_msgq_put(&stream->in_queue, &buffer, K_NO_WAIT);
+				if (ret != 0) {
+					LOG_ERR("buffer %p -> in_queue %p err %d",
+					buffer, &stream->in_queue, ret);
+				}
+			}
+		} else {
+			/* Received a STOP/DRAIN trigger */
+			i2s_mcux_rx_stream_disable(dev, true);
+			stream->state = I2S_STATE_READY;
+		}
+		break;
+	case I2S_STATE_ERROR:
+		i2s_mcux_rx_stream_disable(dev, true);
+		break;
+	}
+}
+
+static int i2s_mcux_tx_stream_start(const struct device *dev)
+{
+	int ret = 0;
+	void *buffer;
+	const struct i2s_mcux_config *cfg = dev->config;
+	struct i2s_mcux_data *dev_data = dev->data;
+	struct stream *stream = &dev_data->tx;
+	I2S_Type *base = cfg->base;
+
+	/* retrieve buffer from input queue */
+	ret = k_msgq_get(&stream->in_queue, &buffer, K_NO_WAIT);
+	if (ret != 0) {
+		LOG_ERR("No buffer in input queue to start transmission");
+		return ret;
+	}
+
+	i2s_mcux_config_dma_blocks(dev, I2S_DIR_TX, (uint32_t *)buffer);
+
+	/* put buffer in output queue */
+	ret = k_msgq_put(&stream->out_queue, &buffer, K_NO_WAIT);
+	if (ret != 0) {
+		LOG_ERR("failed to put buffer in output queue");
+		return ret;
+	}
+
+	/* Enable TX DMA */
+	base->FIFOCFG |= I2S_FIFOCFG_DMATX_MASK;
+
+	ret = dma_start(stream->dev_dma, stream->channel);
+	if (ret < 0) {
+		LOG_ERR("dma_start failed (%d)", ret);
+		return ret;
+	}
+
+	I2S_Enable(base);
+	I2S_EnableInterrupts(base, (uint32_t)kI2S_TxErrorFlag);
+
+	return 0;
+}
+
+static int i2s_mcux_rx_stream_start(const struct device *dev)
+{
+	int ret = 0;
+	void *buffer[NUM_DMA_BLOCKS];
+	const struct i2s_mcux_config *cfg = dev->config;
+	struct i2s_mcux_data *dev_data = dev->data;
+	struct stream *stream = &dev_data->rx;
+	I2S_Type *base = cfg->base;
+	uint8_t num_of_bufs;
+
+	num_of_bufs = k_mem_slab_num_free_get(stream->cfg.mem_slab);
+
+	/*
+	 * Need at least two bffers on the RX memory slab for
+	 * reliable DMA reception.
+	 */
+	if (num_of_bufs == 1) {
+		return -EINVAL;
+	}
+
+	ret = k_mem_slab_alloc(stream->cfg.mem_slab, &buffer[0],
+						K_NO_WAIT);
+	if (ret != 0) {
+		LOG_ERR("buffer alloc from mem_slab failed (%d)", ret);
+		return ret;
+	}
+
+	/*
+	 * Allocate a second receive buffer from SLAB. This will allow
+	 * us to use a Ping-Pong buffer scheme during receive
+	 */
+	ret = k_mem_slab_alloc(stream->cfg.mem_slab, &buffer[1],
+						K_NO_WAIT);
+	if (ret != 0) {
+		LOG_ERR("buffer alloc from mem_slab failed (%d)", ret);
+		return ret;
+	}
```

Comment:
Let's maybe make a small for loop out of this. The code will be easier to read and it might be smaller.

Reply:
Sounds good.